src/wayland/buffer/mod.rs

```diff
@@ -0,0 +1,183 @@
+    pub fn from_wl(buffer: &WlBuffer, dh: &mut DisplayHandle<'_>) -> Buffer {
+        match dh.get_object_data(buffer.id()) {
+            Ok(data) => {
+                match data.downcast::<BufferData>() {
+                    Ok(data) => Buffer(BufferInner::Managed {
+                        buffer: buffer.id(),
+                        data,
+                    }),
+
+                    // The buffer has some user data but is not managed by Smithay.
+                    Err(_) => Buffer(BufferInner::Unmanaged(buffer.clone())),
+                }
+            }
+
+            // A completely unmanaged buffer (generally EGL)
+            Err(_) => Buffer(BufferInner::Unmanaged(buffer.clone())),
+        }
+    }
+
+    /// Returns the object id of the underlying [`WlBuffer`].
+    pub fn id(&self) -> ObjectId {
+        match &self.0 {
+            BufferInner::Managed { buffer, .. } => buffer.clone(),
+            BufferInner::Unmanaged(buffer) => buffer.id(),
+        }
+    }
+
+    /// Returns a reference to the underlying [`WlBuffer`].
+    pub fn buffer(&self, dh: &mut DisplayHandle<'_>) -> Result<WlBuffer, InvalidId> {
+        match &self.0 {
+            BufferInner::Managed { buffer, .. } => WlBuffer::from_id(dh, buffer.clone()),
+            BufferInner::Unmanaged(buffer) => Ok(buffer.clone()),
+        }
+    }
+
+    /// Sends a `release` event, indicating the buffer is no longer in use by the compositor, meaning the
+    /// client is free to reuse or destroy the buffer.
+    pub fn release(&self, dh: &mut DisplayHandle<'_>) -> Result<(), InvalidId> {
+        self.buffer(dh)?.release(dh);
+        Ok(())
+    }
+
+    /// Returns the data associated with the buffer.
+    ///
+    /// This function is intended for abstractions to obtain the buffer data they store. Users should use the
+    /// functions provided by the buffer abstractions to obtain the data. For example, data associated with an
+    /// shm buffer should be obtained using [`with_buffer_contents`](crate::wayland::shm::with_buffer_contents)
+    /// instead of specifying the type of the data in the `T` generic for this function.
+    pub fn buffer_data<T>(&self) -> Result<Option<&T>, UnmanagedResource>
+    where
+        T: Send + Sync + 'static,
+    {
+        match self.0 {
+            BufferInner::Managed { ref data, .. } => Ok(<dyn Any>::downcast_ref::<T>(&*data.data)),
+            BufferInner::Unmanaged(_) => Err(UnmanagedResource),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+enum BufferInner {
+    /// The buffer is managed by Smithay.
+    Managed {
+        /// Object id of the [`WlBuffer`].
+        ///
+        /// This cannot be a [`WlBuffer`] or else there will be an Arc cycle.
+        buffer: ObjectId,
+
+        /// Data associated with the managed buffer.
+        data: Arc<BufferData>,
+    },
+
+    /// Buffer not managed by Smithay.
+    ///
+    /// There is no user data that may be taken from the buffer.
+    Unmanaged(WlBuffer),
+}
```

Comment:
I'm wondering if the inner fields of this enum should be public, or if we should wrap it into a unit struct so that `Buffer` effectively is an opaque type from the point of view of the user.

Reply:
Not sure it's worth exposing the inner fields. Okay the user has an object ID and buffer data. It's not exactly clear what to do with the returned data.

If this were to be public, then I'd need to make more details of BufferData public.

Plus I have added functions to get the data in side these enum variants anyways.

Reply:
Ok, right, my bad, for some reason I though this was actually the public type. :shrug: All is good then.